pkg/operator/storageclasscontroller/storageclasscontroller.go

```diff
@@ -41,147 +27,117 @@ type StorageClassController struct {
+	infra := apiDeps.GetInfrastructure()
 
-	policyName, err := c.syncStoragePolicy(ctx)
+	apiClient, err := newStoragePolicyAPI(ctx, connection, infra)
 	if err != nil {
-		klog.Errorf("error syncing storage policy: %v", err)
-		return err
+		reason := fmt.Errorf("error connecting to vcenter API: %v", err)
+		return "", checks.MakeGenericVCenterAPIError(reason)
 	}
 
-	err = c.syncStorageClass(ctx, syncContext, policyName)
-	if err != nil {
-		klog.Errorf("error syncing storage class: %v", err)
-		return err
-	}
-	_, _, err = v1helpers.UpdateStatus(
-		ctx,
-		c.operatorClient,
-		v1helpers.UpdateConditionFn(controllerAvailableCondition),
-	)
-	return err
-}
+	// we expect all API calls to finish within apiTimeout or else operator might be stuck
+	tctx, cancel := context.WithTimeout(ctx, apiTimeout)
+	defer cancel()
 
-func (c *StorageClassController) syncStoragePolicy(ctx context.Context) (string, error) {
-	infra, err := c.infraLister.Get(infraGlobalName)
+	policyName, err := apiClient.createStoragePolicy(tctx)
 	if err != nil {
-		return "", err
+		return "", checks.MakeGenericVCenterAPIError(err)
 	}
+	return policyName, checks.MakeClusterCheckResultPass()
+}
 
-	cloudConfig := infra.Spec.CloudConfig
-	cloudConfigMap, err := c.configMapLister.ConfigMaps(cloudConfigNamespace).Get(cloudConfig.Name)
-	if err != nil {
-		return "", fmt.Errorf("failed to get cloud config: %v", err)
+func (c *StorageClassController) updateConditions(ctx context.Context, lastCheckResult checks.ClusterCheckResult, clusterStatus checks.ClusterCheckStatus) error {
+	availableCnd := operatorapi.OperatorCondition{
+		Type:   c.name + operatorapi.OperatorStatusTypeAvailable,
+		Status: operatorapi.ConditionTrue,
 	}
 
-	cfgString, ok := cloudConfigMap.Data[infra.Spec.CloudConfig.Key]
-	if !ok {
-		return "", fmt.Errorf("cloud config %s/%s does not contain key %q", cloudConfigNamespace, cloudConfig.Name, cloudConfig.Key)
+	degradedCondition := operatorapi.OperatorCondition{
+		Type:   c.name + operatorapi.OperatorStatusTypeDegraded,
+		Status: operatorapi.ConditionFalse,
 	}
-
-	cfg := new(vsphere.VSphereConfig)
-	err = gcfg.ReadStringInto(cfg, cfgString)
-	if err != nil {
-		return "", err
+	allowUpgradeCond := operatorapi.OperatorCondition{
+		Type:   c.name + operatorapi.OperatorStatusTypeUpgradeable,
+		Status: operatorapi.ConditionTrue,
 	}
 
-	username, password, err := c.getCredentials(cfg)
-	if err != nil {
-		return "", fmt.Errorf("failed to get credentials: %v", err)
+	switch clusterStatus {
+	case checks.ClusterCheckDegrade:
+		degradedMessage := fmt.Sprintf("Marking cluster degraded because %s", lastCheckResult.Reason)
+		klog.Warningf(degradedMessage)
+		c.recorder.Warningf(string(lastCheckResult.CheckStatus), "Marking cluster degraded because %s", lastCheckResult.Reason)
+		degradedCondition.Status = operatorapi.ConditionTrue
+		degradedCondition.Message = degradedMessage
```

Comment:
Only `lastCheckResult.Reason` should be in the condition message, "Marking cluster degraded because %s" is implied by the condition name.

(Applies to all conditions below)

Reply:
fixed.